wlb: T7977: Fix weight calculation for multiple interfaces

l0crian1 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->
When more than 2 interfaces were configured in a WLB rule, nftables would fail to load the config since there were duplicate keys in the vmap. This corrects that issue.
## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
https://vyos.dev/T7977
## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## How to test / Smoketest result
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```

Or provide the output of the smoketest

```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
#### Configure 3 or more interfaces in a WLB rule:
```
set load-balancing wan rule 10 interface eth0.20
set load-balancing wan rule 10 interface eth0.101
set load-balancing wan rule 10 interface eth0.102
```
### Before:
#### Check for presence of nft chain:
```
sudo nft list table vyos_wanloadbalance
Error: No such file or directory
list table vyos_wanloadbalance
           ^^^^^^^^^^^^^^^^^^^
```
#### Errors can be viewed in the log. You'll see errors like this:
```
/run/nftables_wlb.conf:15:139-139: Error: Could not process rule: File exists
        iifname "eth1" ct state new counter numgen random mod 3 vmap { 0 : jump wlb_mangle_isp_eth0.20, 1 : jump wlb_mangle_isp_eth0.101, 1 : jump wlb_mangle_isp_eth0.102 }
                                                                                                                                          ^
```
### After:
#### Check the nft chain. vmap keys are now correct:
```
sudo nft list table vyos_wanloadbalance
table ip vyos_wanloadbalance {
        chain wlb_nat_postrouting {
                type nat hook postrouting priority srcnat - 1; policy accept;
                ct mark 0x000000c9 oifname "eth0.20" counter packets 0 bytes 0 snat to 192.168.2.186
                ct mark 0x000000ca oifname "eth0.101" counter packets 0 bytes 0 snat to 10.0.101.207
                ct mark 0x000000cb oifname "eth2" counter packets 0 bytes 0 snat to 10.0.101.208
        }

        chain wlb_mangle_prerouting {
                type filter hook prerouting priority mangle; policy accept;
                iifname "eth1" ct state new counter packets 0 bytes 0 numgen random mod 12 vmap { 0 : jump wlb_mangle_isp_eth0.20, 1 : jump wlb_mangle_isp_eth2, 2-11 : jump wlb_mangle_isp_eth0.101 }
                iifname "eth1" counter packets 0 bytes 0 meta mark set ct mark
        }

        chain wlb_mangle_output {
                type filter hook output priority mangle; policy accept;
                meta mark != 0x00000000 counter packets 0 bytes 0 return
                meta l4proto icmp counter packets 198 bytes 16632 return
                ip saddr 127.0.0.0/8 ip daddr 127.0.0.0/8 counter packets 24 bytes 1440 return
                oifname != "eth1" ct state new counter packets 0 bytes 0 numgen random mod 12 vmap { 0 : jump wlb_mangle_isp_eth0.20, 1 : jump wlb_mangle_isp_eth2, 2-11 : jump wlb_mangle_isp_eth0.101 }
                oifname != "eth1" counter packets 0 bytes 0 meta mark set ct mark
        }

        chain wlb_mangle_isp_eth0.20 {
                meta mark set 0x000000c9 ct mark set 0x000000c9 counter packets 0 bytes 0 accept
        }

        chain wlb_mangle_isp_eth0.101 {
                meta mark set 0x000000ca ct mark set 0x000000ca counter packets 0 bytes 0 accept
        }

        chain wlb_mangle_isp_eth2 {
                meta mark set 0x000000cb ct mark set 0x000000cb counter packets 0 bytes 0 accept
        }
}
```
### Smoketest results:
```
test_3_or_more_interfaces_in_rule (__main__.TestLoadBalancingWan.test_3_or_more_interfaces_in_rule) ... ok
test_check_chains (__main__.TestLoadBalancingWan.test_check_chains) ... ok
test_criteria_failover_hook (__main__.TestLoadBalancingWan.test_criteria_failover_hook) ... ok
test_firewall_groups (__main__.TestLoadBalancingWan.test_firewall_groups) ... ok
test_table_routes (__main__.TestLoadBalancingWan.test_table_routes) ... ok

----------------------------------------------------------------------
Ran 5 tests in 142.330s

OK
```
## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

sever-sever commented:
Provided example confusing
```
set load-balancing wan rule 10 interface eth0.20
set load-balancing wan rule 10 interface eth0.101
set load-balancing wan rule 10 interface eth0.102
```
What does `eth2` do here?
```
table ip vyos_wanloadbalance {
        chain wlb_nat_postrouting {
                type nat hook postrouting priority srcnat - 1; policy accept;
                ct mark 0x000000c9 oifname "eth0.20" counter packets 0 bytes 0 snat to 192.168.2.186
                ct mark 0x000000ca oifname "eth0.101" counter packets 0 bytes 0 snat to 10.0.101.207
                ct mark 0x000000cb oifname "eth2" counter packets 0 bytes 0 snat to 10.0.101.208
        }
```

l0crian1 commented:
@sever-sever the output was just what I had in my nftables at the time since I needed to make sure both equal weight and unequal weight configs work. 

What was in the output wasn’t really important, just that it existed after the change when 3 or more interfaces were configured.

c-po reviewed (approved):
Fix issue in non auto-incrementing weight for multiple interfaces.

Could you extend the smoketests to validate this szenario?

l0crian1 commented:
> Could you extend the smoketests to validate this szenario?

I considered doing that, but I worried it could be fragile. For instance, when you first configure WLB, it may not fail since the interfaces are not in an active state (the vmap would be empty at that point). So it may pass no matter what. It's only when the health-checks pass and the interfaces go into an active state would cause illegal nftables config.

sever-sever commented:
Smoke tests passed health checks, as I remember. We use vethpairs https://github.com/vyos/vyos-1x/blob/89e4e09ece8e54303f3d5e1f1ab4927e2351b24a/smoketest/scripts/cli/test_load-balancing_wan.py#L170

l0crian1 commented:
Yeah, but it's being done with a bunch of static sleep timers to overcome that fragility.  I may have a better way for that, so it's not adding a bunch of time to the testing.

l0crian1 commented:
@c-po @sever-sever I added a smoketest along with a helper function that waits for a condition to be met or reach a timeout.

l0crian1 committed:
wlb: T7977: Fix weight calculation for multiple interfaces

- Fixed issue in T7977
- Added smoketest for 3 or more interfaces in rule
- Added wait_for function to utils/misc.py

sarthurdev reviewed (approved):
Good catch and thanks for the fix. Tested locally.

Do you think it'd be beneficial to also verify the verdict map in a smoketest?

l0crian1 commented:
> Do you think it'd be beneficial to also verify the verdict map in a smoketest?

The accuracy of the vmaps appears to be checked in `test_check_chains`, so I think that would already be covered.

sarthurdev commented:
> > Do you think it'd be beneficial to also verify the verdict map in a smoketest?
> 
> The accuracy of the vmaps appears to be checked in `test_check_chains`, so I think that would already be covered.

I was thinking more to extend it so it also covers instances of more than 2 interfaces.

l0crian1 commented:
There's probably some value to that. A lot of users with multiple interfaces may just do this when they have 2 interfaces:
```
set load-balancing wan rule 10 interface eth0
set load-balancing wan rule 10 interface eth1 weight 2
```
But with 3 or more interfaces, they may have weights across multiple interfaces (or all of them), which makes the logic a little more rigid. It seems to be handled fine right now, but it could be good regression protection to have it covered in a smoketest.

l0crian1 commented:
@sarthurdev Do you want me to extend the smoketest I added to validate the vmaps have the correct number of weight buckets? Or would it be added in a different PR?

c-po commented:
@l0crian1 please extend the smoketest and then we can merge it.

l0crian1 committed:
wlb: T7977: Updated smoketest to validate nft vmap weight buckets

l0crian1 commented:
@c-po Done!
